### ISSUE TITLE
update '/metrics/delete' to support deleting tagged metrics

### ISSUE THREAD

**Issue by Dieterbe** [grafana/metrictank#1023]
https://graphite.readthedocs.io/en/latest/tags.html#removing-series-from-the-tagdb
graphite has this endpoint because you manually delete series, then issue this request to get the series deleted from the tag index.

for metrictank, i think it doesn't make sense to have that endpoint, because you delete series via /metrics/delete which calls Idx.Delete() which currently doesn't clean up the tag index, but should, in which case '/tags/delSeries'  is no longer needed

**Comment by replay** [comment]
i think we should leave the `/tags/delSeries` endpoint, because otherwise we're not compatible with the graphite API https://graphite.readthedocs.io/en/latest/tags.html#removing-series-from-the-tagdb

we could extend the normal delete endpoint so it understands `seriesByTags()`. allowing the user to delete tagged metrics by their name tag, as if they were untagged metrics would be very inconsistent with the querying side, but allowing the user to use `seriesByTags()` would make sense.

**Comment by gelldur** [comment]
Currently there is no possibility to remove tagged series.

**Comment by shanson7** [comment]
We have processes in place that use this endpoint. Removing it would break those.

**Comment by Dieterbe** [comment]
does /tags/delSeries currently delete the tagged metric from *all* the relevant data structures? Do we agree that it should?

**Comment by replay** [comment]
yes and yes

**Comment by Dieterbe** [comment]
OK so then I think the only thing left to do here is update /metrics/delete to be equivalent to /tags/delSeries, right?

**Comment by replay** [comment]
well, not equivalent. But I'd just extend it so it understands the `seriesByTags()` function, just like the render endpoint. then `/metrics/delete` can also be used to delete tagged metrics

**Comment by fitzoh** [comment]
#1656 is open which updates `/metrics/delete` to recognize `seriesByTags` queries, but I'm doing a bit of second-guessing...

I'm a bit concerned that this might make it a bit too easy to accidentally delete more metrics than intended.

**Comment by replay** [comment]
After some more discussions today we decided that it would be too risky to implement `seriesByTag()` on the `/metrics/delete` endpoint. This would allow a user to accidentally delete all metrics in their index, so it's better to not implement that.
If we really want to provide a way to delete tagged metrics from the `/metrics/delete` endpoint, even though we already have another endpoint for that, then we'd better come up with a different way of querying those metrics to be deleted. But that function should then not be named `seriesByTag()` because that would be confusing if the `seriesByTag()` on the delete endpoint behaves differently than the one on the query endpoint.

**Comment by replay** [comment]
I'd like to a come to a conclusion regarding the question of what we actually want our delete endpoints to look like. There's been some discussion in this issue and also in #1658
Let's just stick to this issue for the discussion because otherwise it's hard to follow everything.

Trying to sum up all the things which I think we need to consider to come to a conclusion
1) @shanson7 needs an endpoint which allows deletions of tagged metrics based on the normal `seriesByTag()` evaluation logic which does not exclude those metrics that have additional un-queried tags
2) we don't want a user who's familiar with graphite's `/tags/delSeries` endpoint to assume MT works the same way and then get surprised because more series have been deleted than intended
3) for consistency we should have a `seriesByTag()` function which works the same way as on the querying side. that way a user can first submit a `/render` request using `seriesByTag()` to look at the metrics which are going to get deleted and then copy-paste that `seriesByTag()` with the parameters into the delete endpoint and rely on that exactly those metrics are going to get deleted.
4) we don't want users to accidentally wipe their index by doing something like a delete of `seriesByTag("name!=")` (matching everything)

Based on these points I think the best solution would still be to have these two ways of deleting tagged metrics:

* `/tags/delSeries` works the same way as in Graphite. meaning that if someone deletes `a.b.c;d=e` then this series gets deleted but `a.b.c;d=e;f=g` would not get deleted. this is mainly because of `2)`
* `/metrics/delete` has a `seriesByTag()` function which works the same way as the one in the `/render` targets. this way `1)` and `3)` are satisfied. 

regarding the idea to add a flag to `seriesByTag()` to make it change it's behavior: I think that's unnecessarily complicated. If we want it to optionally only select those metrics which don't have more tags then the ones which have been queried for, then we also need to decide how we deal with operators other than `=` because in these cases it might be less obvious how that rule gets applied, then we end up defining a whole new query language for that optional more restrictive mode. 

in our cloud user facing docs I simply wouldn't even mention the `/tags/delSeries` endpoint and only talk about the `/metrics/delete` one and make sure it's clear how it works so nobody accidentally deletes more than intended.

This solution ^^ unfortunately does not satisfy `4)`. To some degree I'd just say that with great power comes great responsibility, and we should give the users as much power to do what they want to do as possible. We just need to be sure that the behavior is documented well.

**Comment by shanson7** [comment]
`tags/findSeries` already supports `seriesByTag`-like functionality with `expressions`
